File: bmmo/misc.py
import sys
import json
from bmmo.constants import MAP_OBJECTS_INDEX, BLOCK_SIZE, BLOCKS_AND_WALL_NAMES, WALL_TOOL, BLOCKS
from bmmo.parsing import get_bmap_measures, separate_objects_from_other_objects, from_blocks_to_list_of_blocks_table
from bmmo.optimization import get_least_blocks_from_blocks_table
from bmmo.block import Block


def string_to_json(string: str) -> dict:
    _json = json.loads(string)
    return _json


def print_table(table: list[list[int]]):
    for row in table:
        print(row)


def merge_map_objects_to_string(blocks: list[Block], scale: int, others: list) -> str:
    final_string = ""
    bmap = {}

    last_index = len(others)

    def create_object_from_block(block: Block, id: int) -> dict:
        return {
            "X": str(block.x * scale),
            "Y": str(block.y * scale),
            "ObjWallWidth": str(block.width * scale),
            "ObjWallHeight": str(block.height * scale),
            "ID": str(id),
            "ObjIndexID": "46",
            "Name": "Wall Tool",
            "LogidID": str(id),
            "Poly": "0",
            "ObjIsTile": "0",
            "Depth": "500",
            "ObjType": block.type,
            "ObjSound": block.sound,
            "Team": "-1",
        }

    for index, obj in enumerate(others):
        if 'Author' in obj:
            bmap["Config"] = obj
        else:
            bmap[f"OBJ{index}"] = obj

    for index, block in enumerate(blocks):
        block_id = last_index + index
        bmap[f"OBJ{block_id}"] = create_object_from_block(block, block_id)

    has_crashed = False
    try:
        final_string = json.dumps(bmap)
    except:
        has_crashed = True
        print('fail to serialize via "json.dumps"')

    if not has_crashed:
        return final_string, has_crashed
    else:
        return bmap, has_crashed


def optimize(og_lines: list[str]):
    """Takes a list of lines that represents a bmap.txt.
    Returns a list of lines that represents the newly optimized bmap.txt
    """
    # parse the file
    original_lines = og_lines
    bmap_json = string_to_json(original_lines[MAP_OBJECTS_INDEX])
    _map_width, _map_height = get_bmap_measures(bmap_json)

    # put the blocks and walls in a single list
    blocks_to_optimize, objects_to_keep = separate_objects_from_other_objects(bmap_json, BLOCKS_AND_WALL_NAMES)

    # create blocks table from blocks/walls objects of the map
    list_of_blocks_table = from_blocks_to_list_of_blocks_table(
        _map_width, _map_height,
        blocks_to_optimize,
        WALL_TOOL, BLOCKS, BLOCK_SIZE
    )

    list_of_blocks = []

    # optimize the blocks
    for blocks_table in list_of_blocks_table:
        curr_blocks = get_least_blocks_from_blocks_table(blocks_table)
        list_of_blocks = [*list_of_blocks, *curr_blocks]

    # log some information about the optimization
    len_original = len(blocks_to_optimize)
    len_optimized = len(list_of_blocks)
    compression_ratio = int((1 - (len_optimized / len_original)) * 100)
    print(f"Before: {len_original}, After: {len_optimized}")
    print(f"Compression ratio: {compression_ratio}%")
    if not (len_optimized < len_original):
        print("No optimization to be made")
        return

    # create a list of lines representing a new optimized bmap.txt
    optimized_objects_string, has_crashed = merge_map_objects_to_string(list_of_blocks, BLOCK_SIZE, objects_to_keep)

    stats = {
        "before": len_original,
        "after": len_optimized,
        "ratio": compression_ratio,
        "has_crashed": has_crashed,
    }

    optimized_lines = []
    for line in original_lines[0:MAP_OBJECTS_INDEX]:
        optimized_lines.append(line)
    optimized_lines.append(optimized_objects_string)
    optimized_lines.append(original_lines[MAP_OBJECTS_INDEX + 1])
    return optimized_lines, stats

File: src/bmmo/misc.ts
import {
  MAP_OBJECTS_INDEX,
  BLOCK_SIZE,
  BLOCKS_AND_WALL_NAMES,
  WALL_TOOL,
  BLOCKS,
} from "./constants";
import {
  getBmapMeasures,
  separateObjectsFromOtherObjects,
  fromBlocksToListOfBlocksTable,
} from "./parsing";
import { getLeastBlocksFromBlocksTable } from "./optimization";
import { Block } from "./block";

function stringToJson(str: string): Record<string, any> {
  return JSON.parse(str);
}

function printTable(table: number[][]): void {
  table.forEach((row) => console.log(row.join(" "), "\n"));
}

/**
 * Merges blocks and other map objects into a JSON string.
 * @param {Block[]} blocks - List of blocks to merge.
 * @param {number} scale - Scale to apply to block coordinates and dimensions.
 * @param {any[]} others - List of other map objects to include.
 * @returns {[string, boolean]} The resulting JSON string and a crash flag.
 */
export function mergeMapObjectsToString(
  blocks: Block[],
  scale: number,
  others: any[]
): [string, boolean] {
  const bmap: Record<string, any> = {};
  let hasCrashed = false;

  const createObjectFromBlock = (
    block: Block,
    id: number
  ): Record<string, string> => ({
    X: (block.x * scale).toString(),
    Y: (block.y * scale).toString(),
    ObjWallWidth: (block.width * scale).toString(),
    ObjWallHeight: (block.height * scale).toString(),
    ID: id.toString(),
    ObjIndexID: "46",
    Name: "Wall Tool",
    LogidID: id.toString(),
    Poly: "0",
    ObjIsTile: "0",
    Depth: "500",
    ObjType: block.type,
    ObjSound: block.sound,
    Team: "-1",
  });

  others.forEach((obj, index) => {
    if ("Author" in obj) {
      bmap["Config"] = obj;
    } else {
      bmap[`OBJ${index}`] = obj;
    }
  });

  blocks.forEach((block, index) => {
    const blockId = others.length + index;
    bmap[`OBJ${blockId}`] = createObjectFromBlock(block, blockId);
  });

  let finalString = "";
  try {
    finalString = JSON.stringify(bmap);
  } catch {
    hasCrashed = true;
    console.error('Failed to serialize via "JSON.stringify"');
  }

  return hasCrashed
    ? [JSON.stringify(bmap), hasCrashed]
    : [finalString, hasCrashed];
}

/**
 * Optimizes a map represented as lines of a bmap.txt file.
 * @param {string[]} ogLines - The original lines that represents a bmap.txt.
 * @returns {[string[], Record<string, number | boolean>]} The optimized lines and statistics.
 */
export function optimize(
  ogLines: string[]
): [string[], Record<string, number | boolean>] {
  const originalLines = ogLines;
  const bmapJson = stringToJson(originalLines[MAP_OBJECTS_INDEX]);
  const [_mapWidth, _mapHeight] = getBmapMeasures(bmapJson);

  const [blocksToOptimize, objectsToKeep] = separateObjectsFromOtherObjects(
    bmapJson,
    BLOCKS_AND_WALL_NAMES
  );

  const listOfBlocksTable = fromBlocksToListOfBlocksTable(
    _mapWidth,
    _mapHeight,
    blocksToOptimize,
    WALL_TOOL,
    BLOCKS,
    BLOCK_SIZE
  );

  let listOfBlocks: Block[] = [];

  listOfBlocksTable.forEach((blocksTable) => {
    const currBlocks = getLeastBlocksFromBlocksTable(blocksTable);
    listOfBlocks = [...listOfBlocks, ...currBlocks];
  });

  const lenOriginal = blocksToOptimize.length;
  const lenOptimized = listOfBlocks.length;
  const compressionRatio = Math.floor((1 - lenOptimized / lenOriginal) * 100);

  console.log(`Before: ${lenOriginal}, After: ${lenOptimized}`);
  console.log(`Compression ratio: ${compressionRatio}%`);

  if (lenOptimized >= lenOriginal) {
    console.log("No optimization to be made");
    return [
      ogLines,
      { before: lenOriginal, after: lenOriginal, ratio: 0, hasCrashed: false },
    ];
  }

  const [optimizedObjectsString, hasCrashed] = mergeMapObjectsToString(
    listOfBlocks,
    BLOCK_SIZE,
    objectsToKeep
  );

  const stats = {
    before: lenOriginal,
    after: lenOptimized,
    ratio: compressionRatio,
    hasCrashed,
  };

  const optimizedLines: string[] = [
    ...originalLines.slice(0, MAP_OBJECTS_INDEX),
    optimizedObjectsString,
    originalLines[MAP_OBJECTS_INDEX + 1],
  ];

  return [optimizedLines, stats];
}
